lib/rex/proto/http/web_socket/amazon_ssm.rb

```diff
@@ -0,0 +1,316 @@
+# -*- coding: binary -*-
+
+require 'bindata'
+
+module Rex::Proto::Http::WebSocket::AmazonSsm
+  module PayloadType
+    Output               = 1
+    Error                = 2
+    Size                 = 3
+    Parameter            = 4
+    HandshakeRequest     = 5
+    HandshakeResponse    = 6
+    HandshakeComplete    = 7
+    EncChallengeRequest  = 8
+    EncChallengeResponse = 9
+    Flag                 = 10
+
+    def self.from_val(v)
+      self.constants.find {|c| self.const_get(c) == v }
+    end
+  end
+
+  module UUID
+    def self.unpack(bbuf)
+      sbuf = ""
+      [8...12].each do |idx|
+        sbuf << Rex::Text.to_hex(bbuf[idx])
+      end
+      sbuf << '-'
+      [12...14].each do |idx|
+        sbuf << Rex::Text.to_hex(bbuf[idx])
+      end
+      sbuf << '-'
+      [14...16].each do |idx|
+        sbuf << Rex::Text.to_hex(bbuf[idx])
+      end
+      sbuf << '-'
+      [0...2].each do |idx|
+        sbuf << Rex::Text.to_hex(bbuf[idx])
+      end
+      sbuf << '-'
+      [2...8].each do |idx|
+        sbuf << Rex::Text.to_hex(bbuf[idx])
+      end
+      sbuf.gsub("\\x",'')
+    end
+
+    def self.pack(sbuf)
+      parts = sbuf.split('-').map do |seg|
+        seg.chars.each_slice(2).map {|e| "\\x#{e.join}"}.join
+      end
+      [3, 4, 0, 1, 2].map do |part|
+        Rex::Text.hex_to_raw(parts[part])
+      end.join
+    end
```

Comment:
Are you certain the ordering in here is correct? The reason I ask is there's no reference to any documentation on where it came from and what I usually see is the ordering that Microsoft uses for their packed UUIDs. If the MS packed UUID is the one that's in use, then we could remove this entire class and just use the existing `uuid` datatype provided in [RubySMB::Dcerpc::Uuid](https://github.com/rapid7/ruby_smb/blob/master/lib/ruby_smb/dcerpc/uuid.rb#L5).

The two do not pack the values in the same order however:

```
[4] pry(#<Msf::Modules::Auxiliary__Cloud__Aws__Enum_ssm::MetasploitModule>)> message_id = UUID.rand
=> "dd15ef9d-7d2a-8112-a0d0-c9d42f1e2d23"
[5] pry(#<Msf::Modules::Auxiliary__Cloud__Aws__Enum_ssm::MetasploitModule>)> message_id
=> "dd15ef9d-7d2a-8112-a0d0-c9d42f1e2d23"
[6] pry(#<Msf::Modules::Auxiliary__Cloud__Aws__Enum_ssm::MetasploitModule>)> UUID.pack(message_id)
=> "\xA0\xD0\xC9\xD4/\x1E-#\xDD\x15\xEF\x9D}*\x81\x12"
[7] pry(#<Msf::Modules::Auxiliary__Cloud__Aws__Enum_ssm::MetasploitModule>)> RubySMB::Dcerpc::Uuid.read(UUID.pack(message_id))
=> "d4c9d0a0-1e2f-232d-dd15-ef9d7d2a8112"
```

Reply:
Yeah, their UUID implementation is different (i actually tried using ours at the start - doesnt fly). Check out any of the existing SSM libs - they all have custom UUID packers